pkg/component/controller/metrics.go

```diff
@@ -0,0 +1,359 @@
+/*
+Copyright 2022 k0s authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package controller
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"fmt"
+	"k8s.io/client-go/rest"
+	"net/http"
+	"os"
+	"path"
+	"time"
+
+	"github.com/k0sproject/k0s/internal/pkg/templatewriter"
+	"github.com/k0sproject/k0s/pkg/apis/k0s.k0sproject.io/v1beta1"
+	"github.com/k0sproject/k0s/pkg/component"
+	"github.com/k0sproject/k0s/pkg/constant"
+	"github.com/k0sproject/k0s/pkg/kubernetes"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	namespace       = "k0s-system"
+	pushGatewayName = "k0s-pushgateway"
+)
+
+// Metrics is the reconciler implementation for metrics server
+type Metrics struct {
+	hostname      string
+	log           *logrus.Entry
+	clusterConfig *v1beta1.ClusterConfig
+	tickerDone    context.CancelFunc
+	K0sVars       constant.CfgVars
+	jobs          []*job
+	saver         manifestsSaver
+	restClient    rest.Interface
+}
+
+var _ component.Component = &Metrics{}
+var _ component.ReconcilerComponent = &Metrics{}
+
+// NewMetrics creates new Metrics reconciler
+func NewMetrics(k0sVars constant.CfgVars, saver manifestsSaver, clientCF kubernetes.ClientFactoryInterface) (*Metrics, error) {
+	hostname, err := os.Hostname()
+	if err != nil {
+		return nil, err
+	}
+	log := logrus.WithFields(logrus.Fields{"component": "Metrics"})
+
+	restClient, err := clientCF.GetRESTClient()
+	if err != nil {
+		return nil, fmt.Errorf("error getting REST client for metrics: %w", err)
+	}
+
+	return &Metrics{
+		hostname:   hostname,
+		log:        log,
+		K0sVars:    k0sVars,
+		saver:      saver,
+		restClient: restClient,
+	}, nil
+}
+
+// Init does nothing
+func (m *Metrics) Init(_ context.Context) error {
+	tw := templatewriter.TemplateWriter{
+		Name:     "pushgateway-with-ttl",
+		Template: pushGatewayTemplate,
+		Data: map[string]string{
+			"Namespace": namespace,
+			"Name":      pushGatewayName,
+		},
+	}
+	output := bytes.NewBuffer([]byte{})
+	err := tw.WriteToBuffer(output)
+	if err != nil {
+		return err
+	}
+	err = m.saver.Save("pushgateway.yaml", output.Bytes())
+	if err != nil {
+		return err
+	}
+
+	var j *job
+	j, err = m.newJob("kube-scheduler", "https://localhost:10259/metrics")
+	if err != nil {
+		return err
+	}
+	m.jobs = append(m.jobs, j)
+
+	j, err = m.newJob("kube-controller-manager", "https://localhost:10257/metrics")
+	if err != nil {
+		return err
+	}
+	m.jobs = append(m.jobs, j)
```

Comment:
I think we should also have a job for api-server. Or should we rely that users prom will pull the metrics via `kubernetes` services endpoints?

Reply:
It depends on how we see this feature. If it just exposes cluster components' metrics that can't be scraped in another way, I don't think we need to add the api-server, since end-users can do it themselves.

Reply:
Fair enough